src/ip-validation.ts

```diff
@@ -0,0 +1,341 @@
+  }
+
+  // Handle different trustProxy values
+  if (trustProxy === true) {
+    // Trust all proxies - use leftmost (original client) IP
+    return normalizeIP(forwardedIPs[0])
+  }
+
+  if (typeof trustProxy === 'number') {
+    // Trust N proxies - skip N IPs from the right
+    const index = Math.max(0, forwardedIPs.length - trustProxy)
+    return normalizeIP(forwardedIPs[index])
+  }
+
+  if (typeof trustProxy === 'string') {
+    // Trust specific IP/subnet - check if socket IP matches
+    // For loopback, linklocal, uniquelocal, or specific IPs
+    const trustedRanges: Record<string, string[]> = {
+      loopback: ['127.0.0.0/8', '::1/128'],
+      linklocal: ['169.254.0.0/16', 'fe80::/10'],
+      uniquelocal: ['10.0.0.0/8', '172.16.0.0/12', '192.168.0.0/16', 'fc00::/7']
+    }
+
+    const rangesToCheck = trustedRanges[trustProxy] || [trustProxy]
+    const isTrusted = socketIP
+      ? rangesToCheck.some((range) => {
+          // Handle both single IPs and CIDR notation
+          const cidr = range.includes('/') ? range : `${range}/32`
+          return isIPInRange(socketIP, cidr)
+        })
+      : false
+
+    if (isTrusted) {
+      return normalizeIP(forwardedIPs[0])
+    }
+  }
+
+  // Default to socket address if trust conditions not met
+  return socketIP
+}
+
+/**
+ * Validate a list of IP addresses/CIDR ranges.
+ * Returns an array of error messages (empty if all valid).
+ */
+export function validateIPList(ips: string[]): string[] {
+  const errors: string[] = []
+  for (const ip of ips) {
+    const trimmed = ip.trim()
+    if (!trimmed) continue
+
+    if (trimmed.includes('/')) {
+      // CIDR notation
+      if (!parseCIDR(trimmed)) {
+        errors.push(`Invalid CIDR: ${trimmed}`)
+      }
+    } else {
+      // Single IP
+      if (parseIPv4(trimmed) === null && parseIPv6(trimmed) === null) {
+        errors.push(`Invalid IP: ${trimmed}`)
+      }
+    }
+  }
+  return errors
+}
+
+/**
+ * Create Express middleware that filters requests by source IP.
+ * Uses req.ip which respects Express's trustProxy setting.
+ */
+export function createIPFilterMiddleware(
+  getConfig: () => { allowedSourceIPs?: string[] }
+): RequestHandler {
+  return (req: Request, res: Response, next: NextFunction) => {
+    // Use req.ip which respects Express's trustProxy setting
+    // This ensures X-Forwarded-For is only trusted when trustProxy is configured
+    const ip = req.ip ? normalizeIP(req.ip) : undefined
+    const config = getConfig()
+
+    if (!isIPAllowed(ip, config.allowedSourceIPs)) {
+      debug('Blocked request from %s to %s %s', ip, req.method, req.path)
+      res.status(403).json({
+        state: 'DENIED',
+        statusCode: 403,
+        message: 'Request not allowed from this IP address'
+      })
+      return
+    }
+    next()
+  }
+}
+
+/**
+ * Check if an IP address is allowed for socket-level filtering.
+ * Convenience wrapper for use with raw TCP sockets.
+ */
+export function isSocketIPAllowed(
```

Comment:
It looks like this calls isIPAllowed that calls isIPInRange that calls normalizeIP, so is not needed at all.